### PR TITLE
[FIRRTL] Full Migration to LowerAnnotations

### PR DESCRIPTION
**Pull request by seldridge**
This completes the migration from Annotation scattering/application from the FIRRTL parser into the `LowerAnnotations` pass. This is doing four things, broken up into separate commits:

1. All missing annotation definitions are added. I grabbed these from one internal design. (It is worthwhile to check other designs and make sure nothing was missed.)
2. The resolution/application for all annotations is added to `LowerAnnotations` table.
3. The FIRRTL Parser now uses the "raw annotations" path exclusively.
4. All dead functions are removed.

Closes #2170 (as it supersedes it).

### CODE REVIEW

**Review by dtzSiFive — APPROVED** [review]
It's happening!!!

LGTM, seems fairly straightforward now that all the groundwork is in place.  Awesome work!
(I can't speak to the semantics for all of these annotations, but the ones I'm familiar with look good)

* Love the separate commits and messages for each, :+1: and thanks! (although I'm always a little sad when squash drops the nice explanations :smiling_face_with_tear:)
* Awesome that moving to LowerAnnotations is "just" adding to the table.  I know this was the plan and remember the discussions but still it's neat that that's all that's needed to move them over :grin:!

Added some minor comments, but nothing functional :smile:.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 1]
```diff
@@ -3138,12 +3138,9 @@ struct FIRCircuitParser : public FIRParser {
...
 private:
-  /// Add annotations from a string to the internal annotation map.  Report
-  /// errors using a provided source manager location and with a provided error
-  /// message
-  ParseResult importAnnotations(CircuitOp circuit, SMLoc loc,
-                                StringRef circuitTarget,
-                                StringRef annotationsStr, size_t &nlaNumber);
+  /// Extract Annotations from a JSON-encoded Annotation array string and add
+  /// them to a vector of attributes.  Report errors using a provided source
+  /// manager location and with a provided error message
```
[the remark this answers is not in the document]
Comment (seldridge) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp:3143]: Copy-pasta. Will fix in a rebase.

File: lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp [hunk 2]
```diff
@@ -279,6 +286,11 @@ struct AnnoRecord {
                                    ApplyState &)>
       applier;
 };
+
+/// Resolution and Application for a firrtl.NoTargetAnnotation.
```
[the remark this answers is not in the document]
Comment (seldridge) [reply lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp:290]: Good idea! I'll take care of this in the rebase.

File: lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp [hunk 3]
```diff
@@ -315,8 +329,49 @@ static const llvm::StringMap<AnnoRecord> annotationRecords{{
     // OMIR Annotations
     {omirAnnoClass, {noResolve, applyOMIR}},
     {omirTrackerAnnoClass, {stdResolve, applyWithoutTarget<true>}},
+    {omirFileAnnoClass, NoTargetAnnotation},
     // Miscellaneous Annotations
-    {dontTouchAnnoClass, {stdResolve, applyDontTouch}}
+    {dontTouchAnnoClass, {stdResolve, applyDontTouch}},
+    {prefixModulesAnnoClass,
+     {stdResolve,
+      applyWithoutTarget<true, FModuleOp, FExtModuleOp, InstanceOp>}},
```
[the remark this answers is not in the document]
Comment (seldridge) [reply lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp:337]: Yeah, this stuff is super slick and is way stricter than SFC checking.  I've made zero modifications to @darthscsi code where he added this.  The only thing I've found this doesn't capture well is restricting application to ports, but not modules.  That could be handled with a template argument to the resolvers.

File: lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp [hunk 4]
```diff
@@ -315,8 +329,49 @@ static const llvm::StringMap<AnnoRecord> annotationRecords{{
...
+    {extractSeqMemsAnnoClass, NoTargetAnnotation},
+    {injectDUTHierarchyAnnoClass, NoTargetAnnotation},
+    {convertMemToRegOfVecAnnoClass, NoTargetAnnotation},
+    {sitestBlackBoxAnnoClass, NoTargetAnnotation},
+    {enumComponentAnnoClass, {noResolve, drop}},
+    {enumDefAnnoClass, {noResolve, drop}},
+    {enumVecAnnoClass, {noResolve, drop}},
+    {forceNameAnnoClass,
+     {stdResolve, applyWithoutTarget<true, FModuleOp, FExtModuleOp>}},
+    {flattenAnnoClass, {stdResolve, applyWithoutTarget<false, FModuleOp>}},
```
[the remark this answers is not in the document]
Comment (seldridge) [reply lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp:348]: I was intending to not perturb this much in this PR and then to clean it up later. One thing is that I am not very happy with how `AnnotationDetails.h` is organized right now and I think that may help with clarity. I have considered something that exactly mirrors the Chisel/Scala class hierarchy with a namespace hierarchy:

```c++
namespace firrtl {
  namespace transforms {
    const char *DontTouchAnnotation = "DontTouchAnnotation";
  }
}
namespace sifive {
  namespace enterprise {
    namespace grandcentral {
      const char *ViewAnnotation = "ViewAnnotation";
    }
  }
}

AnnotationSet annotations(foo);
foo.hasAnnotation(transforms::DontTouchAnnotation);
```

This is likely anal retentive. However, I think it's actually an anti-pattern to invent a name for each of these things, e.g., `testBenchDirAnnoClass` is useless without examining `AnnotationDetails.h`.

File: lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (seldridge) [reply lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp:337]: One other comment is that there are frequent usages of similar groupings of template arguments which correspond to equivalent Scala things. E.g., there is a `firrtl.annotations.IsModule` trait which means "module or instance". This effectively means `#define FModuleOp, FExtModuleOp, InstanceOp`. However, I wasn't sure if there was a clean way of doing this other than with a `#define`. Any thoughts on whether this matters or if there is some clever way to do this?

**Review by youngar — APPROVED** [review]
LGTM!

**Comment by seldridge** [comment]
> Love the separate commits and messages for each, 👍 and thanks! (although I'm always a little sad when squash drops the nice explanations 🥲)

Opted for the rebase-and-merge as I think it was warranted here (and I triple-checked that all commits were passing).

**Pull request merged**